lib/src/embedded/isolate_dispatcher.dart

```diff
@@ -158,19 +158,27 @@ class IsolateDispatcher {
     var receivePort = ReceivePort();
     isolate.sink.add((receivePort.sendPort, compilationId));
 
-    var channel = IsolateChannel<Uint8List?>.connectReceive(receivePort)
-        .transform(const ExplicitCloseTransformer());
-    channel.stream.listen(_channel.sink.add,
+    var channel = IsolateChannel<Uint8List>.connectReceive(receivePort);
+    channel.stream.listen(
+        (message) {
+          // The first byte of messages from isolates indicates whether the
+          // entire compilation is finished. Sending this as part of the message
+          // buffer rather than a separate message avoids a race condition where
+          // the host might send a new compilation request with the same ID as
+          // one that just finished before the [IsolateDispatcher] receives word
+          // that the isolate with that ID is done. See sass/dart-sass#2004.
+          _channel.sink.add(Uint8List.sublistView(message, 1));
```

Comment:
To be extra safe, this should come after the if branch below. In other words, after isolate is deactivated / resource is released.

Reply:
This has to come _before_ `_channel.sink.close()`, or the call to `add()` will fail. I think it's safe enough to rely on Dart's single-threading guarantees here.